File: core/test-fixtures/src/commonMain/kotlin/com/mockdonalds/app/core/test/StateRobot.kt
package com.mockdonalds.app.core.test

abstract class StateRobot<State, Event> {
    private val _capturedEvents = mutableListOf<Event>()
    val capturedEvents: List<Event> get() = _capturedEvents.toList()
    val lastEvent: Event? get() = _capturedEvents.lastOrNull()

    protected fun createEventSink(): (Event) -> Unit = { event ->
        _capturedEvents.add(event)
    }

    fun clearEvents() {
        _capturedEvents.clear()
    }

    abstract fun defaultState(): State
}

File: iosApp/iosAppTests/Unit/StateRobot.swift
import ComposeApp

protocol StateRobot {
    associatedtype State
    associatedtype Event: AnyObject

    var capturedEvents: [Event] { get }
    var lastEvent: Event? { get }
    func createEventSink() -> (Event) -> Void
    func defaultState() -> State
    func clearEvents()
}

class BaseStateRobot<State, Event: AnyObject>: StateRobot {
    private(set) var capturedEvents: [Event] = []
    var lastEvent: Event? { capturedEvents.last }

    func createEventSink() -> (Event) -> Void {
        { [weak self] event in self?.capturedEvents.append(event) }
    }

    func clearEvents() { capturedEvents.removeAll() }

    func defaultState() -> State {
        fatalError("Subclass must override defaultState()")
    }
}
